### ISSUE TITLE
Can't reproduce results for YouCookII

### ISSUE THREAD

**Issue by dzabraev** [antoine77340/S3D_HowTo100M#2]
I took this model
```
wget https://www.rocq.inria.fr/cluster-willow/amiech/howto100m/s3d_howto100m.pth
wget https://www.rocq.inria.fr/cluster-willow/amiech/howto100m/s3d_dict.npy
```
and code from this repository, I take validation part of youcookII and try to achieve numbers
mentioned in the article [End-to-End Learning of Visual Representations from Uncurated Instructional Videos](https://arxiv.org/pdf/1912.06430.pdf)

![Capture](https://user-images.githubusercontent.com/15128988/80914706-0c393100-8d56-11ea-99af-918a21c74736.PNG)

It is unclear which protocol did you use for testing. In the following table I show several experiments and none of them could  achieve your results. Could you clarify which test protocol did you use for testing? It will be good if you publish script for testing.

What I try.

* `T` is time in seconds. I split each clip to subclips each has length T seconds. For each subclip embedding will be compute.
* `pooling` If clip  was split to >1 subclips embeddings will be averaged to `pooling`
* `imgsz` Short side of each source video will be rescaled to `imgsz` with h:w preserving. Then
center crop will be taken for each frame.
* `normalize` Whether or not sentence embedding and each video embedding was L2-normalized before dot-product.
* `num frames` From each `T`-seconds clip `num frames` was taken in uniform style.
* `num resample`  For each clip sample different `num resample` sets of frames. For each resample compute embedding. With `pooling` all embedding will be polled to single one. LCR means sample from each clip 3 times: `num frames` left crops, `num_frames` right crops, `num frames` center crops.

| T   |imgsize|pooling|normalize|num frames|num resample|   R@1|   R@5|R@10|MedR|
|----|---------|--------|-----------|-------------|----------------|-------|-------|-------|-------|
|250| 200      |max     |False        |32               | 1                    |11.478 |27.610 |37.453 |21|
|250| 224      |max     |False        |32               | 1                    |8.774 | 22.044  | 30.975 | 32 |
|250| 256      |max     |False        |32               | 1                    |5.912 | 15.503 | 21.038 |104 |
|1.5| 200      |max     |False        |32               | 1                     |8.333 |23.208 |31.981 |31|
|3.2| 200      |max     |False        |32               | 1                     |9.497 | 24.969 | 34.654 | 24 |
|8| 200      |max     |False        |32               | 1                        |10.094 |25.818 |35.849 |23|
|16| 200      |max     |False        |32               | 1                      |10.755 |26.478 |36.541 |21|
|32| 200      |max     |False        |32               | 1                      |11.164 |27.484 |37.296|21|
|64| 200      |max     |False        |32               | 1                      |11.415 |27.704 |37.547 |21|
|128| 200      |max     |False        |32               | 1                     |11.447|27.610 |37.453 |21|
|250| 200      |max     |True        |32               | 1                     |9.906 |25.031 |34.748 |25|
|250| 200      |max    |False        |32               | 2                     |11.604 |28.270 |37.987 |20|
|250| 200      |max    |False        |32               | 3                     |11.918 |28.396 |38.333 |21|
|250| 200      |max    |False        |32               | 4                     |11.509 | 28.082  |38.365 |21|
|250| 200      |max    |False        |32               | LCR                | 11.384 |27.138 |37.704 |22 |
|250| 200      |mean|False        |32               | 4                     |12.075 |28.805 |38.459 |20|

**Comment by antoine77340** [comment]
Hi,

Thank you for your comment here is how we have done it:
Each testing video is sampled at 10 fps and rescaled so that min(height,width) = 224.
For each YouCook2 video clip, we sampled 5 x 32-frame clips linearly spaced (so each clip is of 3.2 seconds), center crop it to 224x224, and compute the video embedding for each of them of size 512. And then we average pool the embeddings.
Finally there was no normalization.
I guess the main difference with what you are doing is that you are uniformly sampling the 32 frames over the whole video right ? Or are the 32 frames sampled, always subsequent frames ?

Also please make sure to put the model in eval mode, otherwise you will recompute batch norm statistics over running batches.

One thing to note is that this pytorch model is a port of the official tensorflow release model from: https://tfhub.dev/deepmind/mil-nce/s3d/1
I did convert the weights to pytorch and did run a benchmark on CrossTask to check if the numbers were similar but I did not check on YouCook2. If you still happen to have any problem, please let me know and I will check myself on YouCook2.

**Comment by dzabraev** [comment]
Do you L2-normalize text embeddings and video embeddings before avg-pooling?
If not, is it ok that text embedding has L2-norm ~175 and video embedding ~0.25 ?

**Comment by antoine77340** [comment]
No normalization is needed, I managed to rerun the YouCook2 evaluation using this pytorch model on a new code (different from my codebase at Deepmind) and with a validation set sligtly larger than the one I had at deepmind and got 49,5 in R@10.
I assume there is a problem in how you sample the video clips of 32 frames. Are they always 32 contiguous frames ? if not then you might have some issues if you randomly sampled 32 frames within a large clip of 250 seconds.

**Comment by dzabraev** [comment]
Thank you for explanation. I succeed to get numbers from article. The main reason was in JPG compression. By default ffmpeg uses JPG compression when it doing unpacking video to images. I disabled compression and could manage to get required number.

**Comment by ra1ph2** [comment]
I am also trying to replicate the results mentioned in the paper on YouCook2 dataset using the pretrained model provided but there seems to be some error that I am not able to locate since I am getting pretty bad results on a random 256 video clip sample of the validation set of YouCook2 validation set. I did the following steps -

- Downloaded the YouCook2 dataset and then divided the untrimmed videos into clips according to the annotated segments durations.
- Randomly sampled a 256 clip validation set for testing the model.
- Loaded the clips as tensors using framerate as 10 and minimum size as 224 and center cropped them.
- Preprocessed the tensors by normalizing them.
- Uniformly selected 5 x 32 frames subclips if there are greater than 5 subclips otherwise took all the subclips and computed the 512 size embeddings and then mean pooled them.
- Computed the text embeddings beforehand and then filled a similiarity matrix for to calculate metrics.

This is the main evaluation code.

```python
def collate_fn(batch):
    vids = [item['video'] for item in batch]
    sents = [item['text'] for item in batch]
    video_id = [item['input'] for item in batch]
    return {'video': vids, 'text': sents, 'input': video_id}

df = pd.read_csv('./YouCook2/YC2_validation_clips_256.csv')
dataset = YouCookClipLoader(
            csv='./YouCook2/YC2_validation_clips_256.csv',
            framerate=10,
            size=224,
            centercrop=1
          )

n_dataset = len(dataset)
loader = DataLoader(
    dataset,
    batch_size=1,
    shuffle=False,
    num_workers=0,
    sampler=None,
    collate_fn=collate_fn
)
preprocess = Preprocessing('3d')

video_net = S3D('s3d_dict.npy', 512)
video_net.load_state_dict(th.load('s3d_howto100m.pth'))
video_net.eval()

loss_op = MaxMarginRankingLoss(
    margin=1,
    negative_weighting=False,
    batch_size=256,
    n_pair=1,
    hard_negative_rate=0.5,
)
loss_op.cuda()

sim_matrix = np.zeros((n_dataset, n_dataset))

video_net.cpu()
text_common_embed = []
with th.no_grad():
    for i, row in df.iterrows():
        video_path = row['video_id']
        name = video_path.split('.')
        name = name[-2]
        name = name.split('/')
        name = name[-1]
        num = name.split('_')
        num = num[-1]
        name = name[:len(name)-len(num)-1]
        video_id = name
        
        sent = youcook_json[video_id]['annotations'][int(num)-1]['sentence']
        sent = [sent]
        text_embed = video_net.text_module(sent)
        # text_embed = F.normalize(text_embed['text_embedding'], p=2, dim=1)
        text_common_embed.append({'embed':text_embed.detach(), 'sent':sent})
        
video_net.cuda()
with th.no_grad():
    for i_batch, sample_batch in enumerate(loader):
        torch.cuda.empty_cache()
        video = sample_batch['video']
        sent = sample_batch['text']
        print("Load: ",len(sent))
        print('Video Embedding Start')
      
        batch_size = len(video)
        video_embed = []
        for i in range(0, batch_size):
            clip = video[i]
            clip = preprocess(clip, 32, 224)

            p = clip
            seq_len = list(p.size())[0]
            if seq_len > 5:
                skip = seq_len // 5
                arr = []
                for j in range(0, seq_len, skip):
                    arr.append(clip[j].unsqueeze(0))
                p = torch.cat(arr, 0)
            vid_len = list(p.size())[0]
            p = p.cuda()
            print(p.shape)
            out = video_net(p)

#             temp,_ = torch.max(out['video_embedding'] , dim=0)
            
            temp = torch.mean(out['video_embedding'] , dim=0)
            temp = temp.unsqueeze(0)        
            video_embed.append(temp.detach().cpu())
        
        video_embed = torch.cat(video_embed, dim=0)
        video_embed = F.normalize(video_embed, p=2, dim=1)
        sentences = sent
        
        for j, text_dict in enumerate(text_common_embed):
            sim_matrix[j,i_batch] = th.matmul(text_embed, video_embed.t()).cpu().numpy()
       
loss = loss_op(th.from_numpy(sim_matrix))
print("Validation Loss: ", loss.item())

metrics = compute_metrics(sim_matrix)
print_computed_metrics(metrics)
```

Can you please point out any obvious errors that are leading to very low retrieval scores. 
@dzabraev It would very helpful if you can provide your testing script to replicate the results.

**Comment by xiangyh9988** [comment]
> I disabled compression and could manage to get required number

Hi bro, sorry to bother you. Could you please share how to disable the JPG compression in ffmpeg-python? I try to search for the arguments but didn't find how to disable it.

**Comment by dzabraev** [comment]
1. Add -q:v 1 to ffmpeg arguments.
2. You can unpack video to bmp. BMP is lossless format. It will give you the best possible quality, but each image will have big size.

**Comment by xiangyh9988** [comment]
Got it. Thank you. After seeing your another issue, I see that you used ffmpeg command-line to unpack videos. I misunderstood that and thought the compression need to be disabled in `ffmpeg-python`. My bad.
